pkg/validation/internal/annotations.go

```diff
@@ -0,0 +1,65 @@
+package internal
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
+	"github.com/operator-framework/api/pkg/validation/errors"
+)
+
+// CaseSensitiveAnnotationKeySet is a set of annotation keys that are case sensitive
+// and can be used for validation purposes. This may not be an exhaustive list.
+var CaseSensitiveAnnotationKeySet = map[string]struct{}{
+
+	v1.OperatorGroupAnnotationKey:             {},
+	v1.OperatorGroupNamespaceAnnotationKey:    {},
+	v1.OperatorGroupTargetsAnnotationKey:      {},
+	v1.OperatorGroupProvidedAPIsAnnotationKey: {},
+	"olm.skipRange":                           {},
+
+	// below are duplicate values of known keys
+
+	// v1alpha1.OperatorGroupNamespaceAnnotationKey:    {},
+	// v1alpha2.OperatorGroupAnnotationKey:             {},
+	// v1alpha2.OperatorGroupNamespaceAnnotationKey:    {},
+	// v1alpha2.OperatorGroupTargetsAnnotationKey:      {},
+	// v1alpha2.OperatorGroupProvidedAPIsAnnotationKey: {},
+}
+
+/*
+ValidateAnnotationNames will check annotation keys to ensure they are using
+proper case. Uses CaseSensitiveAnnotationKeySet as a source for keys
+which are known to be case sensitive. This function can be used anywhere
+annotations need to be checked for case sensitivity.
+
+Arguments
+
+• annotations: annotations map usually obtained from ObjectMeta.GetAnnotations()
+
+• value: is the field or file that caused an error or warning
+
+Returns
+
+Any errors that may have been detected with the annotation keys provided
+*/
+func ValidateAnnotationNames(annotations map[string]string, value interface{}) (errs []errors.Error) {
+	// for every annotation provided
+	for annotationKey := range annotations {
+		// loop through all well known keys
+		for knownCaseSensitiveKey := range CaseSensitiveAnnotationKeySet {
+			// case insensitively compare this annotation key with this specific well known key.
+			// we want to determine if the key was intended to be used
+			if strings.EqualFold(annotationKey, knownCaseSensitiveKey) {
+				// we have a case-insensitive match... now check to see if the case is really correct
+				if annotationKey != knownCaseSensitiveKey {
+					// annotation key supplied is invalid due to bad case.
+					errs = append(errs, errors.ErrFailedValidation(fmt.Sprintf("provided annotation %s uses wrong case and should be %s instead", annotationKey, knownCaseSensitiveKey), value))
+				}
+				// bail out... we found and checked what we're looking for with this annotationKey
+				break
```

Comment:
Should we check every annotations and project the errors for all of them instead of bailing out as soon as we find one problem?

Reply:
I will be reimplementing this given your other comment, but I'll add a response just so I've addressed this comment:

The `break` in this case would only affect the inner loop where we are ranging on `CaseSensitiveAnnotationKeySet` and only when we've established that `annotationKey` matches case insensitively with the `knownCaseSensitiveKey`. In other words, the current `annotationKey` **is** a known case sensitive key (but we don't yet know if it matches in a case sensitive manner yet), so there's no point looking at any other known keys. The remaining annotations provided would continue to be checked when the outer loop continues on.